Add a callback type to enforce message ordering

kamek-pf opened the pull request:
Hey,

This PR adds a `OrderedCallback` to enforce message ordering.
It works like `SimpleCallback`, but instead of running each action to completion in their own thread, it uses a `TQueue` of actions that is consumed in order by a separate thread.

I've been using this in production for a few days, it works as expected.
I used [this repo](https://github.com/kamek-pf/mqtt-ordering) to try different implementations, validate the behavior and run some benchmarks.
In the end it looks like simply reading elements one by one from a `TQueue` yielded the best results.

The new callback is a little slower, as one would expect.

Let me know what you think :)

queezle42 commented:
Using an unbounded queue for arriving messages will result in unbounded memory consumption in case messages arrive faster than they can be handled. There should be no queue or a bounded queue.

When messages can't be handled fast enough there are various valid strategies. The easiest way would be to just block the network connection. That would lead to back-pressure which *may* terminate the MQTT connection - but that is usually preferable to your application being killed by the OOM killer.

This already is a problem with the current SimpleCallback implementation (messages are queued as Haskell threads), so I think this PR is a strict improvement to the current implementation!

dustin commented:
Thanks for the code (and the review)!  I've not had a chance to look yet, but I should be able to do that today.

I do agree with the problem (and the queueing issues).  My only concern is how well things have been working for me for a while.  :)

kamek-pf commented:
I agree, the only reason I used an unbounded queue is because it mirrors the current behavior.

I could change it to a bounded queue, which would require users to specify a maximum size. 
Although in my experience it's often unclear what size should be used so I personally tend to come up with random numbers until I can do some profiling (... if that ever happens :p).

Alternatively, we could keep this one unbounded and add another callback type which would accept a maximum size and use a bounded queue internally.

queezle42 commented:
I don't see a good reason to have a queue at all, a (T)MVar to pass messages to the worker thread should work just as well. The kernel queue for TCP will buffer incoming data already and it's very unlikely that parsing MQTT messages is enough of a bottleneck to warrant storing parsed messages.

In high-throughput scenarios it's probably better to have a handler in the application that quickly dispatches messages to multiple workers (potentially using multiple worker/topic-related queues), instead of relying on a global message queue in the library.

kamek-pf commented:
Oh I see what you mean, that makes sense.
I'll try to implement it this way in the afternoon and check how it behaves !

dustin commented:
Note that the purpose of the threads isn't so much for queueing as it is for error handling.  The thing I don't want is a bad handler that silently causes the client to hang due to programming error.  For this reason, I modeled it more like an HTTP handler type of thing where a message being handled could divide by zero or something.  In cases where an exception is thrown, I want to make sure the connection can drop and nothing leaks.

kamek-pf commented:
I switched to an MVar, this should address the queuing issue.

queezle42 commented on src/Network/MQTT/Client.hs:
```diff
@@ -538,12 +548,16 @@ sendAndWait c@MQTTClient{..} dt f = do
 subscribe :: MQTTClient -> [(Filter, SubOptions)] -> [Property] -> IO ([Either SubErr QoS], [Property])
-subscribe c@MQTTClient{..} ls props = do
+subscribe c@MQTTClient{_cb} ls props = do
+  when (isOrdered _cb) runCallbackThread
```
This is missing a guard against starting multiple callback threads (subscribe might be called multiple times).

queezle42 commented on src/Network/MQTT/Client.hs:
```diff
@@ -699,4 +713,4 @@ registerCorrelated MQTTClient{_corr} bs cb = modifyTVar' _corr (Map.insert bs cb
 -- | Unregister a callback handler for the given correlated data identifier.
 unregisterCorrelated :: MQTTClient -> BL.ByteString -> STM ()
-unregisterCorrelated MQTTClient{_corr} bs = modifyTVar' _corr (Map.delete bs)
+unregisterCorrelated MQTTClient{_corr} bs = modifyTVar' _corr (Map.delete bs)
```
Missing newline character at end of line

queezle42 commented on net-mqtt.cabal:
```diff
@@ -5,7 +5,7 @@ cabal-version: 1.12
 
 name:           net-mqtt
-version:        0.8.1.1
+version:        0.8.2.0
```
Usually version numbers are assigned by the author when a new version is released and should not be changed in a pull request.

dustin commented:
I finally looked at this and it looks a lot less scary than I was imagining. :)  In particular, it looks very unlikely to cause trouble for existing applications.

I'm grateful for the additional review as well.  Let's make sure only one thread gets started for a consumer and that the lifecycle is clear.  I'm also going to want to get these differences documented before a release.  I think there's more info in the PR than the commit messages, so I'd want that to be remembered.  I can clean a lot of that up, though.

kamek-pf commented:
Thank you both for your feedback !

@dustin I added some documentation to explain the differences between SimpleCallback and OrderedCallback, let me know if it's clear enough.

@queezle42 I implemented the requested changes, using a `TVar Bool` for the guard. 

Still, I'm not quite sure what to do with the callback thread if the user explicitly unsubscribes. 
Currently it's simply `link`ed to the same thread as `subscribe`, but it should probably be terminated when the client gets disconnected, or when the user unsubscribes from everything. 

I could keep the async handle and :
- Use `withAsync` to ensure the callback thread dies with the client
- `cancel` it when no subscriptions remain 

Does that sound right ?

queezle42 commented:
Unsubscribing does not indicate there are no queued incoming messages (or the thread might be still be handling a valid message), so cancelling the thread early is a bad idea.

I'd say the best time to terminate the thread is when a message has been handled and the client is disconnected; cancelling the thread immediately when the client disconnects is probably also an acceptable implementation.

And thank you very much for implementing this properly!

kamek-pf commented:
Ok, so I extracted everything we talked about into two functions.
One that starts the ordered callback thread (if needed, and only once), and one that stops it.

I ended up using `finally` to perform the cleanup after `waitForClient` (mostly because I don't see an obvious way to do it otherwise :p).

Is this the appropriate place to do this ?

kamek-pf commented on src/Network/MQTT/Client.hs:
```diff
@@ -114,7 +114,8 @@ data MQTTClient = MQTTClient {
   , _corr         :: TVar (Map BL.ByteString MessageCallback)
   , _cbM          :: MVar (IO ())
-  , _cbHandle     :: TMVar (Async ())
+  , _hasCbThread  :: TVar Bool
+  , _cbHandle     :: MVar (Async ())
```
It might be possible to get away with just one variable containing the async handle, my last attempt wasn't safe though.

dustin replied:
I think we can move forward with this and I can take a pass at that.  It's not obvious why you'd need more than a `TVar (Maybe (Async ())` here, though it might be tedious to get all that in place.

kamek-pf replied:
Should be pretty straightforward actually, I should've just done that. 
Let me know if you're busy, I don't mind changing it.

dustin replied:
You did some good work here I'm happy to tinker with it a bit.  :)

Testing is a bit hard here, *but* it's relatively easy in my mqttd project, so I might try to see if I can have a meaningful test over there before pushing, but in any case, I'll push a new version today.

Thanks a lot for doing this!

dustin reviewed (approved):
Would you mind if I squashed these commits into one with the PR details (keeping your authorship, of course)?

kamek-pf commented:
Oh yeah you should definitely squash, those intermediate commit messages don't bring anything you want to keep around.

dustin commented:
This was squashed and pushed manually.
